src/cache/algorithms/expr_variable_reordering.cpp

```diff
@@ -0,0 +1,253 @@
+  expr_variable_reordering side2_reordering(arith->side_2);
+
+  side1_reordering.run();
+  side2_reordering.run();
+
+  symbols_vec symbols; // annotates all inner symbols
+  values_vec values;   // annotates all inner values
+
+  // 1. Create a list of symbols and values
+  this->transverse_read_binop(arith, symbols, values);
+
+  // 2. Order it in crescent order
+  sort(symbols.begin(), symbols.end(), [](const auto &lhs, const auto &rhs) {
+    return lhs->get_symbol_name() > rhs->get_symbol_name();
+  });
+
+  // A constant propagation should have been executed prior to this
+  assert(values.size() <= 1);
+
+  // 3. Change all values
+  this->transverse_replace_binop(arith, symbols, values);
+}
+
+void expr_variable_reordering::run_on_relation(expr2tc &expr) noexcept
+{
+  std::shared_ptr<relation_data> relation;
+  relation = std::dynamic_pointer_cast<relation_data>(expr);
+
+  // If relation is false then the cast failed
+  assert(relation);
+
+  expr_variable_reordering side1(relation->side_1);
+  side1.run();
+
+  expr_variable_reordering side2(relation->side_2);
+  side2.run();
+}
+
+inline void expr_variable_reordering::transverse_read_binop(
+  const std::shared_ptr<arith_2ops> op,
+  expr_variable_reordering::symbols_vec &symbols,
+  expr_variable_reordering::values_vec &values)
+{
+  transverse_binop(op, symbols, values, TRANSVERSE_MODE::READ);
+}
+
+inline void expr_variable_reordering::transverse_replace_binop(
+  const std::shared_ptr<arith_2ops> op,
+  expr_variable_reordering::symbols_vec &symbols,
+  expr_variable_reordering::values_vec &values)
+{
+  transverse_binop(op, symbols, values, TRANSVERSE_MODE::REPLACE);
+}
+
+void expr_variable_reordering::transverse_binop(
+  const std::shared_ptr<arith_2ops> op,
+  expr_variable_reordering::symbols_vec &symbols,
+  expr_variable_reordering::values_vec &values,
+  expr_variable_reordering::TRANSVERSE_MODE mode)
+{
+  // Check if LHS is the same binary operation of parent
+  if(op->side_1->expr_id == op->expr_id)
+  {
+    std::shared_ptr<arith_2ops> arith;
+    arith = std::dynamic_pointer_cast<arith_2ops>(op->side_1);
+    transverse_binop(arith, symbols, values, mode);
+  }
+
+  switch(mode)
+  {
+  case TRANSVERSE_MODE::READ:
+  {
+    add_value(op, true, symbols, values);
+    break;
+  }
+  case TRANSVERSE_MODE::REPLACE:
+  {
+    replace_value(op, true, symbols, values);
+    break;
+  }
+  }
+
+  // Check if RHS is the same binary operation of parent
+  if(op->side_2->expr_id == op->expr_id)
+  {
+    std::shared_ptr<arith_2ops> arith;
+    arith = std::dynamic_pointer_cast<arith_2ops>(op->side_2);
+    transverse_binop(arith, symbols, values, mode);
+  }
+
+  switch(mode)
+  {
+  case TRANSVERSE_MODE::READ:
+  {
+    add_value(op, false, symbols, values);
+    break;
+  }
+  case TRANSVERSE_MODE::REPLACE:
+  {
+    replace_value(op, false, symbols, values);
+    break;
+  }
+  }
+}
+
+void expr_variable_reordering::add_value(
```

Comment:
Add short description for all methods.

Reply:
The header includes the description of the method